scene/resources/primitive_meshes.cpp

```diff
@@ -1592,6 +1592,136 @@ bool SphereMesh::get_is_hemisphere() const {
+	Vector<Vector3> normals;
+	Vector<float> tangents;
+	Vector<Vector2> uvs;
+	Vector<int> indices;
+
+#define ADD_TANGENT(m_x, m_y, m_z, m_d) \
+	tangents.push_back(m_x);            \
+	tangents.push_back(m_y);            \
+	tangents.push_back(m_z);            \
+	tangents.push_back(m_d);
+
+	float min_radius = inner_radius;
+	float max_radius = outer_radius;
+
+	if (min_radius == max_radius) {
+		return; //sorry, can't
+	}
+
+	if (min_radius > max_radius) {
+		SWAP(min_radius, max_radius);
+	}
+
+	float radius = (max_radius - min_radius) * 0.5;
+
+	for (int i = 0; i <= rings; i++) {
+		int prevrow = (i - 1) * (ring_segments + 1);
+		int thisrow = i * (ring_segments + 1);
+		float inci = float(i) / rings;
+		float angi = inci * Math_TAU;
+
+		Vector2 normali = Vector2(-Math::sin(angi), -Math::cos(angi));
+
+		for (int j = 0; j <= ring_segments; j++) {
+			float incj = float(j) / ring_segments;
+			float angj = incj * Math_TAU;
+
+			Vector2 normalj = Vector2(-Math::cos(angj), Math::sin(angj));
+			Vector2 normalk = normalj * radius + Vector2(min_radius + radius, 0);
+
+			points.push_back(Vector3(normali.x * normalk.x, normalk.y, normali.y * normalk.x));
+			normals.push_back(Vector3(normali.x * normalj.x, normalj.y, normali.y * normalj.x));
+			ADD_TANGENT(-Math::cos(angi), 0.0, Math::sin(angi), 1.0);
+			uvs.push_back(Vector2(inci, incj));
+
+			if (i > 0 && j > 0) {
+				indices.push_back(thisrow + j - 1);
+				indices.push_back(prevrow + j);
+				indices.push_back(prevrow + j - 1);
+
+				indices.push_back(thisrow + j - 1);
+				indices.push_back(thisrow + j);
+				indices.push_back(prevrow + j);
+			}
+		}
+	}
+
+	p_arr[RS::ARRAY_VERTEX] = points;
+	p_arr[RS::ARRAY_NORMAL] = normals;
+	p_arr[RS::ARRAY_TANGENT] = tangents;
+	p_arr[RS::ARRAY_TEX_UV] = uvs;
+	p_arr[RS::ARRAY_INDEX] = indices;
+}
+
+void TorusMesh::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_inner_radius", "radius"), &TorusMesh::set_inner_radius);
+	ClassDB::bind_method(D_METHOD("get_inner_radius"), &TorusMesh::get_inner_radius);
+
+	ClassDB::bind_method(D_METHOD("set_outer_radius", "radius"), &TorusMesh::set_outer_radius);
+	ClassDB::bind_method(D_METHOD("get_outer_radius"), &TorusMesh::get_outer_radius);
+
+	ClassDB::bind_method(D_METHOD("set_rings", "rings"), &TorusMesh::set_rings);
+	ClassDB::bind_method(D_METHOD("get_rings"), &TorusMesh::get_rings);
+
+	ClassDB::bind_method(D_METHOD("set_ring_segments", "rings"), &TorusMesh::set_ring_segments);
+	ClassDB::bind_method(D_METHOD("get_ring_segments"), &TorusMesh::get_ring_segments);
+
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "inner_radius", PROPERTY_HINT_RANGE, "0.001,1000.0,0.001,or_greater,exp"), "set_inner_radius", "get_inner_radius");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "outer_radius", PROPERTY_HINT_RANGE, "0.001,1000.0,0.001,or_greater,exp"), "set_outer_radius", "get_outer_radius");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "rings", PROPERTY_HINT_RANGE, "3,64,1"), "set_rings", "get_rings");
```

Comment:
For large and thin toruses, having more than 64 rings can be useful to improve detail:

```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::INT, "rings", PROPERTY_HINT_RANGE, "3,128,1"), "set_rings", "get_rings");
```